### ISSUE TITLE
dataflow API

### ISSUE THREAD

**Issue by lionelperrin** [ruby-concurrency/concurrent-ruby#312]
Hello,

Thank you for this library. It really is filling a big gap in the standard ruby library. I don't want to report an issue but would rather discuss about two things:

* Dataflow API. The dataflow API is very convenient for the application I'm developping. I think that it helps keeping dependencies between tasks very clear. Nevertheless, I face an issue due to the exception management. I would like that the arguments given to the block not being the value of the future so that it is possible to write code like this:
```
task_a = Concurrent::dataflow{ puts "first task"; 10 }
b = Concurrent::dataflow(task_a ) do |a|
  a.value # should rethrow the exception from task_a if it has got an exception
          # or return the value from task_a if no exception has occured.
  if  a.rejected? # keep the possibility to handle errors from task_a
    0
  else
    a.value + 10
  end
end
```

* Are there any attempts to develop a cluster using concurrent-ruby ? I have akka-cluster in mind. A big part of the API of concurrent-ruby seems generic enough to be applied to remote threads as well as it is applied to local threads today. This would make possible to write programs that can be seamless executed on a single interpreter or being distributed on several machines. Is it something that can be done ? (I know that there is a real difficulty to execute a ruby block on a remote interpreter due to the fact that blocks can't be serialized.)

**Comment by chrisseaton** [comment]
Maybe we could have a variant of `Concurrent::dataflow`, `Concurrent::dataflow_futures` that gives you the futures instead of their values? Of course you could still just access `task_a` inside your block.

As you say reference clusters, sending a Ruby block across a network connection is problematic. It would probably need to be solved at the level of MRI (and JRuby, and Rubinius...). Also, even if you can send a block, how do you send the rest of the state it needs? All the classes that have been loaded that it might use and so on?

**Comment by pitr-ch** [comment]
It seem that `Concurrent::Promise` might offer what you are looking for.

You can also look at new Future/Promise framework in `concurrent-edge` gem, just released in 0.9.pre. See https://github.com/ruby-concurrency/concurrent-ruby/blob/master/examples/edge_futures.out.rb for examples. it offers better performance an requires less thread to operate. In your case it could look as follow:
```ruby
a = Concurrent.future { raise if rand > 0.5; 10 }
b = a.then { |v| v+10 } # propagates error if a fails
b.value! # returns 20 or raises same error which occurred in a

result = Concurrent.
    future { raise if rand > 0.5; 10 }.
    then { |v| v+10 }. 
    rescue { |reason| 0 }
result.value! # => 20 or 0 on error 

Concurrent.
    future { raise if rand > 0.5; 10 }.
    chain { |success, value, reason| some_custom_handling } 
```

**Comment by lionelperrin** [comment]
Thank you for pointing this out, I like the syntax associated to future chaining. It has a very clear meaning.
In my case, some futures depends on several other. I think this can be solved with the promise API which allows to trigger a computation when all promises from a list are available. It is solved with a very clear API using dataflow as well. Is there any equivalent with future I have missed ?

`Concurrent::dataflow_futures` is more what I was expecting. In the meantime it is true that I can access task_a from within the block. In my case, task_a is returned by a function so my code is more likely to look like below, which is not awful.
```
Concurrent::dataflow(in_a = build_dataflow_a, in_b = build_dataflow_b) do 
  a = in_a.value!
  b = in_b.rejected? 0 : in_b.value
end
```

**Comment by pitr-ch** [comment]
There is a `zip` which allows to connect together many futures:
```ruby
head    = Concurrent.future { 1 } 
branch1 = head.then(&:succ) 
branch2 = head.then(&:succ).then(&:succ) 
# zipping futures
branch1.zip(branch2).value # => [2, 3]
(branch1 & branch2).then { |a, b| a + b }.value! # => 5
(branch1 & branch2).then(&:+).value! # => 5
Concurrent.zip(branch1, branch2, branch1).then { |*values| values.reduce &:+ }.value! # => 7
```

**Comment by lionelperrin** [comment]
Oh great! Look like it is quite easy to replicate the dataflow API using only futures. It looks quite easy to handle simple error cases using `rescue` but it is still difficult to manage cases where only one input is missing.
I understand that this can be solved using a construct like this:
```
branch1 = Concurrent.future{1}
branch2 = Concurrent.future{ raise 'boum' }
(branch1 & branch2.rescue{0}).then{ |a, b| a+b }.value!
```
But in my opinion, writing like this might be easier and more flexible:
```
branch1 = Concurrent.future{1}
branch2 = Concurrent.future{ raise 'boum' }
(branch1 & branch2)._then_{ |a, b| b.rejected? a : a+b }.value!
```

**Comment by pitr-ch** [comment]
It comes from the `then` convention: do the block only on success so on zip on success of all the futures. There is `chain` method however which returns triplet `success, value, reason` this should be modified on `zip(*).chain` to return whole set. Now it behaves too much like then returning `success` only of all the futures were success, `value` array of values, `reason` first failure. I'll modify this. Thanks for your feedback. 

BTW `success, value, reason` is returned to avoid reading from the future itself which is volatile operation so little bit slower.

**Comment by jdantonio** [comment]
@lionelperrin May I close this issue, or do you need additional assistance?

**Comment by pitr-ch** [comment]
zip(*).chain implemented in #349
